include/mockturtle/algorithms/mig_inv_propagation.hpp

```diff
@@ -0,0 +1,293 @@
+        std::swap( a, b );
+    }
+
+    /* trivial cases */
+    if ( a.index == b.index )
+    {
+      return ( a.complement == b.complement ) ? a : c;
+    }
+    else if ( b.index == c.index )
+    {
+      return ( b.complement == c.complement ) ? b : a;
+    }
+
+    std::shared_ptr<storage<regular_node<3, 2, 1>>>::element_type::node_type nd;
+    nd.children[0] = a;
+    nd.children[1] = b;
+    nd.children[2] = c;
+
+    /* structural hashing */
+    const auto it = ntk._storage->hash.find( nd );
+    if ( it != ntk._storage->hash.end() )
+    {
+      return { it->second, 0 };
+    }
+
+    const auto index = ntk._storage->nodes.size();
+
+    if ( index >= .9 * ntk._storage->nodes.capacity() )
+    {
+      ntk._storage->nodes.reserve( static_cast<uint64_t>( 3.1415f * index ) );
+      ntk._storage->hash.reserve( static_cast<uint64_t>( 3.1415f * index ) );
+    }
+
+    ntk._storage->nodes.push_back( nd );
+
+    ntk._storage->hash[nd] = index;
+
+    /* increase ref-count to children */
+    ntk._storage->nodes[a.index].data[0].h1++;
+    ntk._storage->nodes[b.index].data[0].h1++;
+    ntk._storage->nodes[c.index].data[0].h1++;
+
+    for ( auto const& fn : ntk._events->on_add )
+    {
+      ( *fn )( index );
+    }
+
+    return { index, 0 };
+  }
+
+private:
+  Ntk& ntk;
+  mig_inv_propagation_stats& st;
+};
+
+} // namespace detail
+
+/*! \brief MIG inverter propagation.
+ *
+ * This algorithm tries to push all
+ * the inverters to the inputs.
+ * However, it can increase the number
+ * of nodes while doing so.
+ *
+ * **Required network functions:**
+ * get_node
+ * substitute_node
+ * take_out_node
+ * foreach_fanin
+ * replace_in_node
+ * fanout_size
+ * is_complemented
+ * is_dead
+ * is_constant
+ * is_pi
+ * foreach_po
+ */
+template<class Ntk>
+void mig_inv_propagation( Ntk& ntk, mig_inv_propagation_stats* pst = nullptr )
+{
+  static_assert( is_network_type_v<Ntk>, "Ntk is not a network type" );
+  static_assert( has_get_node_v<Ntk>, "Ntk does not implement the get_node method" );
+  static_assert( has_substitute_node_v<Ntk>, "Ntk does not implement the substitute_node method" );
+  static_assert( has_take_out_node_v<Ntk>, "Ntk does not implement the take_out_node method" );
+  static_assert( has_foreach_fanin_v<Ntk>, "Ntk does not implement the foreach_fanin method" );
+  static_assert( has_replace_in_node_v<Ntk>, "Ntk does not implement the replace_in_node method" );
+  static_assert( has_fanout_size_v<Ntk>, "Ntk does not implement the fanout_size method" );
+  static_assert( has_is_complemented_v<Ntk>, "Ntk does not implement the is_complemented method" );
+  static_assert( has_is_dead_v<Ntk>, "Ntk does not implement the is_dead method" );
+  static_assert( has_is_constant_v<Ntk>, "Ntk does not implement the is_constant method" );
+  static_assert( has_is_pi_v<Ntk>, "Ntk does not implement the is_pi method" );
+  static_assert( has_foreach_po_v<Ntk>, "Ntk does not implement the foreach_po method" );
```

Comment:
Shouldn't you add a `static_assert` check for the MIG base type? Based on the code and the original paper, it seems like the algorithm is only applicable to MIGs.